src/common/ad_tx.c

```diff
@@ -608,3 +613,423 @@ ad_tx_redo_act_next(struct ad_tx *tx)
 	}
 	return &tx->tx_redo_act_pos->it_act;
 }
+
+static struct ad_tx *
+tx_get()
+{
+	return tls_tx;
+}
+
+static void
+tx_set(struct ad_tx *tx)
+{
+	tls_tx = tx;
+}
+
+static void
+tx_callback(struct ad_tx *tx)
+{
+	if (!tx->tx_stage_cb || tx->tx_layer != 0)
+		return;
+	tx->tx_stage_cb(ad_tx_stage(tx), tx->tx_stage_cb_arg);
+}
+
+static int
+tx_end(struct ad_tx *tx, int err)
+{
+	struct umem_tx	*utx;
+	int		 rc;
+
+	if (err == 0)
+		err = tx->tx_last_errno;
+	else
+		tx->tx_last_errno = err;
+
+	D_ASSERTF(tx->tx_layer >= 0, "TX "DF_U64", bad layer %d\n", ad_tx_id(tx), tx->tx_layer);
+	if (tx->tx_layer != 0)
+		return 0;
+
+	rc = ad_tx_end(tx, err);
+	if (err == 0) {
+		if (rc == 0) {
+			ad_tx_stage_set(tx, UMEM_STAGE_ONCOMMIT);
+		} else {
+			D_ERROR("ad_tx_end(%d) failed, "DF_RC"\n", err, DP_RC(rc));
+			tx->tx_last_errno = rc;
+			ad_tx_stage_set(tx, UMEM_STAGE_ONABORT);
+		}
+	} else {
+		D_ASSERT(rc != 0);
+		D_ASSERTF(ad_tx_stage(tx) == UMEM_STAGE_ONABORT,
+			  "TX "DF_U64", bad stage %d\n", ad_tx_id(tx), ad_tx_stage(tx));
+	}
+
+	tx_callback(tx);
+	if (tx_get() == tx)
+		tx_set(NULL);
+	/* Give a chance to UMEM_STAGE_NONE callback.
+	 * XXX with less opportunity comparing with PMDK as user may started new TX here,
+	 *     or change to always call it even if with new TX in TLS?
+	 */
+	if (tx_get() == NULL) {
+		ad_tx_stage_set(tx, UMEM_STAGE_NONE);
+		tx_callback(tx);
+	}
+	rc = tx->tx_last_errno;
+	utx = ad_tx2umem_tx(tx);
+	D_FREE(utx);
+
+	return rc;
+}
+
+static int
+tx_abort(struct ad_tx *tx, int err)
+{
+	if (err == 0)
+		err = -DER_CANCELED;
+
+	ad_tx_stage_set(tx, UMEM_STAGE_ONABORT);
+
+	return tx_end(tx, err);
+}
+
+int
+mo_ad_tx_begin(struct umem_instance *umm, struct umem_tx_stage_data *txd)
```

Comment:
I suggest to make this a generic function shared by both AD and BMEM.

Reply:
instead of exporting all these functions, I think we should just export "ad_ops" (please rename to ad_mem_ops), and make all these functions be "static", and remove _ad_  from prefix.
e.g. umo_tx_gein, umo_tx_end.

The only reason we defined pmdk callbacks in mem.c is because we didn't have other plugin at that time